### PR TITLE
Adds SVG rendering capabilities to the main viewer.

### PR DESCRIPTION
**Pull request by yurydelendik**
Currently it will be pref'd off, but can be switched at components to demo viewer app level.

### CODE REVIEW

File: web/pdf_page_view.js [hunk 1]
```diff
@@ -40,6 +40,9 @@ var RenderingStates = pdfRenderingQueue.RenderingStates;
 
 var TEXT_LAYER_RENDER_DELAY = 200; // ms
 
+var CANVAS_RENDERER = 'canvas';
+var SVG_RENDERER = 'svg';
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js]: I quick drive-by comment, but would it perhaps make sense to use an enumeration here instead, similar to many other places in the code? E.g.
```javascript
var Renderer = {
  CANVAS: 0, // Default
  SVG: 1,
};
```
Or it you need/want the content to be `string`s (e.g. for passing that in from `app.js`), it may still be clearer with e.g.
```javascript
var Renderer = {
  CANVAS: 'canvas', // Default
  SVG: 'svg',
};
```

File: web/pdf_page_view.js [hunk 2]
```diff
@@ -519,40 +580,72 @@ var PDFPageView = (function PDFPageViewClosure() {
...
-      }
-      div.setAttribute('data-loaded', true);
+      return result;
+    },
 
-      if (self.onBeforeDraw) {
-        self.onBeforeDraw();
+    paintOnSvg: function PDFPageView_paintOnSvg(canvasWrapper) {
+      if (typeof PDFJSDev !== 'undefined' &&
+          PDFJSDev.test('FIREFOX || MOZCENTRAL || CHROME')) {
```
Comment (timvandermeij) [inline web/pdf_page_view.js]: Indenting is off in these two lines: four and three spaces used instead of two.

File: web/app.js [hunk 3]
```diff
@@ -214,6 +214,7 @@ var PDFViewerApplication = {
       downloadManager: downloadManager,
       enhanceTextSelection: false,
       renderInteractiveForms: false,
+      renderer: 'canvas',
```
Comment (timvandermeij) [inline web/app.js]: Can we make this a viewer preference? It will be easier to work with/toggle it.

File: web/pdf_page_view.js [hunk 4]
```diff
@@ -423,35 +374,36 @@ var PDFPageView = (function PDFPageViewClosure() {
...
         // triggering this callback.
-        if (renderTask === self.renderTask) {
-          self.renderTask = null;
+        if (paintTask === self.paintTask) {
+          self.paintTask = null;
         }
 
         if (error === 'cancelled') {
-          rejectRenderPromise(error);
+          self.error = null;
```
Comment (timvandermeij) [inline web/pdf_page_view.js]: The member variable `error` does not appear to be used anywhere in this file. Can we remove this or am I missing something here?

File: web/pdf_page_view.js [hunk 5]
```diff
@@ -582,6 +603,50 @@ var PDFPageView = (function PDFPageViewClosure() {
 
       return result;
     },
+
+    paintOnSvg: function PDFPageView_paintOnSvg(canvasWrapper) {
+      if (typeof PDFJSDev !== 'undefined' &&
```
Comment (timvandermeij) [inline web/pdf_page_view.js]: If we make this a viewer preference, we may not need this. In that case, we may also remove https://github.com/yurydelendik/pdf.js/blob/4f51d694c1f8792ed98d56f4ef17fb80bd8754ba/src/display/svg.js#L27.

File: web/pdf_page_view.js [hunk 6]
```diff
@@ -582,6 +603,50 @@ var PDFPageView = (function PDFPageViewClosure() {
...
+      var ensureNotCancelled = function () {
+        if (cancelled) {
+          throw 'cancelled';
+        }
+      };
+
+      var self = this;
+      var pdfPage = this.pdfPage;
+      var SVGGraphics = pdfjsLib.SVGGraphics;
+      var viewport = this.viewport;
```
Comment (timvandermeij) [inline web/pdf_page_view.js]: This intermediate variable is not really used. Let's remove it and use `this.viewport` in the line below.

File: web/pdf_page_view.js [hunk 7]
```diff
@@ -582,6 +603,50 @@ var PDFPageView = (function PDFPageViewClosure() {
...
+      var actualSizeViewport = viewport.clone({scale: CSS_UNITS});
+      var promise = pdfPage.getOperatorList().then(function (opList) {
+        ensureNotCancelled();
+        var svgGfx = new SVGGraphics(pdfPage.commonObjs, pdfPage.objs);
+        return svgGfx.getSVG(opList, actualSizeViewport).then(function (svg) {
+          ensureNotCancelled();
+          self.svg = svg;
+          svg._viewport = actualSizeViewport;
```
Comment (timvandermeij) [inline web/pdf_page_view.js]: What does this do?

File: web/pdf_page_view.js (hunk 4 above)
Comment (Snuffleupagus) [inline web/pdf_page_view.js]: https://github.com/mozilla/pdf.js/blob/master/web/app.js#L1609

File: web/pdf_page_view.js (hunk 7 above)
Comment (Snuffleupagus) [inline web/pdf_page_view.js]: Nothing in practice, but I assume it's there simply to maintain parity with the `canvas` code; see https://github.com/mozilla/pdf.js/blob/master/web/pdf_page_view.js#L403-L404 and https://github.com/mozilla/pdf.js/blob/master/web/pdf_page_view.js#L259-L260.

Comment (Snuffleupagus) [inline web/pdf_page_view.js]: See the existing code for the `canvas` case: https://github.com/mozilla/pdf.js/blob/master/web/pdf_page_view.js#L403-L404 and https://github.com/mozilla/pdf.js/blob/master/web/pdf_page_view.js#L259-L260.

File: web/pdf_page_view.js [hunk 8]
```diff
@@ -483,33 +454,123 @@ var PDFPageView = (function PDFPageViewClosure() {
...
         });
+      };
 
-        if (!error) {
-          resolveRenderPromise(undefined);
-        } else {
-          rejectRenderPromise(error);
+      var paintTask = this.renderer === CANVAS_RENDERER ?
+        this.paintOnCanvas(canvasWrapper) :
+        this.paintOnSvg(canvasWrapper);
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js]: Nit: Given that the default is `canvas`, to me it seems a bit more logical to invert this condition:
```javascript
var paintTask = this.renderer === SVG_RENDERER ?
  this.paintOnSvg(canvasWrapper) :
  this.paintOnCanvas(canvasWrapper) :
```

File: web/pdf_page_view.js [hunk 9]
```diff
@@ -189,6 +197,17 @@ var PDFPageView = (function PDFPageViewClosure() {
         rotation: totalRotation
       });
 
+      if (this.svg) {
+        this.cssTransform(this.svg, true);
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js]: Nit: With `cssTransform` now being used with SVG rendering as well, it's a bit confusing that the method signature is still
`cssTransform: function PDFPageView_transform(canvas, redrawAnnotations) {`.

I thus think that it would be clearer if the `canvas` parameter was changed to e.g. `renderTarget` (or similar), so the method signature instead becomes:
```javascript
cssTransform: function PDFPageView_transform(renderTarget,
                                             redrawAnnotations) {
```

File: web/pdf_page_view.js (hunk 8 above)
Comment (timvandermeij) [inline web/pdf_page_view.js]: Moreover, it may be good to rename `canvasWrapper` to `wrapper`, or `pageWrapper` if you prefer, as it's not specific to canvases anymore.

File: web/pdf_page_view.js (hunk 1 above)
Comment (yurydelendik) [inline web/pdf_page_view.js]: RendererType string 'enum' is added to ui_utils.js

File: web/app.js (hunk 3 above)
Comment (yurydelendik) [inline web/app.js]: done as with renderInteractiveForms

**Comment by timvandermeij** [comment]
Awesome! Thank you for working on this.
